src/Lean/Meta/Tactic/Try/Collect.lean

```diff
@@ -0,0 +1,210 @@
+  return ((← getEnv).getModuleIdxFor? declName).isNone
+
+def getFunInductName (declName : Name) : Name :=
+  declName ++ `induct
+
+def getFunInduct? (declName : Name) : MetaM (Option Name) := do
+  let .defnInfo _ ← getConstInfo declName | return none
+  try
+    let result ← realizeGlobalConstNoOverloadCore (getFunInductName declName)
+    return some result
+  catch _ =>
+    return none
+
+def isEligible (declName : Name) : M Bool := do
+  if declName.hasMacroScopes then
+    return false
+  if (← getConfig).main then
+    return (← inCurrentModule declName)
+  if (← getConfig).name then
+    let ns ← getCurrNamespace
+    return ns.isPrefixOf declName
+  return false
+
+def saveEqnCandidate (declName : Name) : M Unit := do
+  if (← isEligible declName) then
+    let some eqns ← getEqnsFor? declName | return ()
+    if eqns.isEmpty then return ()
+    unless (← Grind.isEMatchTheorem eqns[0]!) do
+      modify fun s => { s with eqnCandidates := s.eqnCandidates.insert declName }
+
+def getEqDefDecl? (declName : Name) : MetaM (Option Name) := do
+  let declName := declName ++ `eq_def
+  if (← Grind.isEMatchTheorem declName) then return none
+  try
+    let result ← realizeGlobalConstNoOverloadCore declName
+    return some result
+  catch _ =>
+    return none
+
+def saveUnfoldCandidate (declName : Name) : M Unit := do
+  if (← isEligible declName) then
+    let some eqDefName ← getEqDefDecl? declName | return ()
+    modify fun s => { s with unfoldCandidates := s.unfoldCandidates.insert eqDefName }
+
+def visitConst (declName : Name) : M Unit := do
+  saveConst declName
+  saveUnfoldCandidate declName
+
+-- Horrible temporary hack: compute the mask assuming parameters appear before a variable named `motive`
+-- It assumes major premises appear after variables with name `case?`
+-- It assumes if something is not a parameter, then it is major :(
+-- TODO: save the mask while generating the induction principle.
+def getFunIndMask? (declName : Name) (indDeclName : Name) : MetaM (Option (Array Bool)) := do
+  let info ← getConstInfo declName
+  let indInfo ← getConstInfo indDeclName
+  let (numParams, numMajor) ← forallTelescope indInfo.type fun xs _ => do
+    let mut foundCase := false
+    let mut foundMotive := false
+    let mut numParams : Nat := 0
+    let mut numMajor : Nat := 0
+    for x in xs do
+      let localDecl ← x.fvarId!.getDecl
+      let n := localDecl.userName
+      if n == `motive then
+        foundMotive := true
+      else if !foundMotive then
+        numParams := numParams + 1
+      else if n.isStr && "case".isPrefixOf n.getString! then
+        foundCase := true
+      else if foundCase then
+        numMajor := numMajor + 1
+    return (numParams, numMajor)
+  if numMajor == 0 then return none
+  forallTelescope info.type fun xs _ => do
+    if xs.size != numParams + numMajor then
+      return none
+    return some (mkArray numParams false ++ mkArray numMajor true)
```

Comment:
https://leanprover-community.github.io/mathlib4_docs/Lean/Meta/Tactic/ElimInfo.html#Lean.Meta.ElimInfo should have some of the code to understand the type of the induction principle (position of motive etc)